src/librustc_trans/trans/intrinsic.rs

```diff
@@ -314,17 +314,55 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
         }
         (_, "size_of") => {
             let tp_ty = *substs.types.get(FnSpace, 0);
-            let lltp_ty = type_of::type_of(ccx, tp_ty);
-            C_uint(ccx, machine::llsize_of_alloc(ccx, lltp_ty))
+            size_of_ty(ccx, tp_ty)
         }
         (_, "min_align_of") => {
             let tp_ty = *substs.types.get(FnSpace, 0);
-            C_uint(ccx, type_of::align_of(ccx, tp_ty))
+            min_align_of_ty(ccx, tp_ty)
         }
         (_, "pref_align_of") => {
             let tp_ty = *substs.types.get(FnSpace, 0);
-            let lltp_ty = type_of::type_of(ccx, tp_ty);
-            C_uint(ccx, machine::llalign_of_pref(ccx, lltp_ty))
+            pref_align_of_ty(ccx, tp_ty)
+        }
+        (_, "size_of_val") => {
+            let tp_ty = *substs.types.get(FnSpace, 0);
+            if type_is_sized(tcx, tp_ty) {
+                size_of_ty(ccx, tp_ty)
+            } else {
+                let ptr = ptr_to_fat_ptr_extra(bcx, ccx, tcx, llargs[0],
+                                               call_debug_location);
+                let size = size_of_unsized_val(bcx, ccx, tcx, ptr, tp_ty,
+                                               call_debug_location);
+                //If the size is not aligned to the alignment, then we need to
```

Comment:
Hm, isn't the size always meant to be divisible by alignment?

At least, we recently had a PR fixing a bug where this accidentally wasn't the case https://github.com/rust-lang/rust/pull/20611

(BTW, the convention is to have a space after the `//` in a comment, `// if the size ...`.)